### PR TITLE
KAFKA-18030 Remove old upgrade-system-tests modules

### PR DESCRIPTION
**Pull request by mingdaoy**
https://issues.apache.org/jira/browse/KAFKA-18030

https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=235838251#KIP896:RemoveoldclientprotocolAPIversionsinKafka4.0-Motivation

> We propose setting the new baseline for protocol API versions to Apache Kafka 2.1 (released 4+ years ago on Nov 2018). More concretely, for each protocol API, we will only keep the most recent version supported by Apache Kafka 2.1 as well as any version introduced after Apache Kafka 2.1

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: build.gradle
Comment (chia7712) [inline build.gradle]: please remove  `treams:upgrade-system-tests-0101:test`, `streams:upgrade-system-tests-0102:test` and `streams:upgrade-system-tests-0110:test` as well

**Review by chia7712 — COMMENTED** [review]
@mingdaoy thanks for this patch

File: build.gradle
Comment (mingdaoy) [reply build.gradle]: Just updated the PR based on your feedback! @chia7712

**Comment by mingdaoy** [comment]
Hi maintainers, could you please review and approve this workflow when you have a chance? Thanks!

**Comment by mingdaoy** [comment]
image

image

**Review by chia7712 — COMMENTED** [review]
@mingdaoy could you please cleanup https://github.com/apache/kafka/blob/trunk/gradle/dependencies.gradle#L188 also?

**Comment by mingdaoy** [comment]
I removed L87 as well. @chia7712 
https://github.com/apache/kafka/blob/trunk/gradle/dependencies.gradle#L87

File: gradle/dependencies.gradle
Comment (chia7712) [inline gradle/dependencies.gradle:186]: please remove other unused versions too

File: gradle/dependencies.gradle
Comment (mingdaoy) [inline gradle/dependencies.gradle:87]: @chia7712

File: build.gradle
Comment (FrankYang0529) [inline build.gradle]: Probably, this part can also be removed.

https://github.com/apache/kafka/blob/50c15b94c94fbe8f964703c057963b38100b0bd6/streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java#L282-L322

**Review by FrankYang0529 — COMMENTED** [review]
Hi @mingdaoy, thanks for the PR. Leave a minor comment.

File: build.gradle
Comment (mjsax) [reply build.gradle]: I think we need to keep this.

While AK 4.0 brokers won't support older clients, one could have a 3.x broker, and still upgrade a KS app from these older versions to AK 4.0.

**Comment by mjsax** [comment]
I am not 100% sure if the PR make sense? While AK 4.0 brokers don't support older client versions any longer, Kafka Streams is also compatible to older broker version.

Thus, one could use a 3.x broker, and upgrade Kafka Streams 0.11.0 to 4.0... To be fair, we would only test this upgrade using `trunk` for brokers right now, what would break as the older client could not connect to a `trunk` broker, but I don't think the broker versions actually matters for what we want to test. In the end, we test if we can upgrade KS client side, which is more or less independent of the user broker version.

Thus, I am actually wondering, if we should instead rewrite the corresponding system test, to use broker version 3.9 to test upgrading older KS versions to 4.0+ client side

**Comment by mjsax** [comment]
To be fair, we don't support all older broker version for a while any longer: https://kafka.apache.org/39/documentation/streams/upgrade-guide#streams_api_broker_compat

Thus, we could also cut it down somewhat and remove maybe 0.10.x releases?

**Comment by chia7712** [comment]
> I am not 100% sure if the PR make sense? While AK 4.0 brokers don't support older client versions any longer, Kafka Streams is also compatible to older broker version.

[KIP-896](https://cwiki.apache.org/confluence/display/KAFKA/KIP-896%3A+Remove+old+client+protocol+API+versions+in+Kafka+4.0) states that 'client compatibility system tests will be updated so that 4.0.x clients fail with an UNSUPPORTED_VERSION error when interacting with Apache Kafka brokers that are too old (2.0 and/or 1.x).' It seems to me that Kafka Streams, being part of the client, should also follow this rule. I mean, there wouldn't be any running Streams instances to upgrade, as they'd encounter an unsupported version error and fail before the rolling upgrade when the broker is 4.0+

> Thus, one could use a 3.x broker, and upgrade Kafka Streams 0.11.0 to 4.0... To be fair, we would only test this upgrade using trunk for brokers right now, what would break as the older client could not connect to a trunk broker, but I don't think the broker versions actually matters for what we want to test. In the end, we test if we can upgrade KS client side, which is more or less independent of the user broker version.
> Thus, I am actually wondering, if we should instead rewrite the corresponding system test, to use broker version 3.9 to test upgrading older KS versions to 4.0+ client side

That's a fair question. Personally, upgrading the client from 0.x/1.x to 4.x is a rare case since it's a giant leap 😄. Therefore, simplifying the test matrix and e2e infra is more important to me at the moment. Additionally, the side effect of adding this test case is that the trunk branch would need to retain all old versions for testing, including upgrade-system-tests and distributions. We don't add such test case for consumer upgrades ( [consumer_protocol_migration_test.py](https://github.com/apache/kafka/blob/trunk/tests/kafkatest/tests/client/consumer_protocol_migration_test.py#L22)).

**Comment by chia7712** [comment]
@mjsax I open https://issues.apache.org/jira/browse/KAFKA-18045 to add 0.11, 1.0, 1.1, and 2.0 back for streams upgrade tests.

**Comment by chia7712** [comment]
@mingdaoy Could you please remove  0.10.x  related code instead?

**Comment by ableegoldman** [comment]
FWIW I tend to share Matthias's concerns here -- to be fair I completely missed KIP-896 and should have spoken up during the discussion for it, so maybe it's too late now. But we've always attempted to maintain compatibility with older broker versions for Kafka Streams, and this feels like a huge leap to go from supporting everything above 0.11 to only 2.1+

Furthermore, AFAICT we aren't giving users any kind of heads up or deprecation notice, not even a warning log (as far as I'm aware) before suddenly breaking this compatibility. 

Now all that said, I personally do find it infuriating when people are running Streams against such ancient broker versions. 2.0 was truly a long long time ago, not to mention 1.x or even 0.11. 

But I also acknowledge that this does happen, and breaking these applications relying on older broker versions without any kind of prior notice feels quite unfair to these users :/

**Comment by mjsax** [comment]
It's a somewhat philosophical question, and maybe we should have a discussion on the dev mailing list (or maybe PMC private list?) to discuss what backward compatibility guarantees we want to provide, and write it down somewhere in the docs and/or wiki?

**Comment by chia7712** [comment]
@mjsax @ableegoldman thanks for all your feedback. I have raised a discussion on the private channel.

**Comment by ijuma** [comment]
> But we've always attempted to maintain compatibility with older broker versions for Kafka Streams, and this feels like a huge leap to go from supporting everything above 0.11 to only 2.1+

Why is it a huge leap? This means removing support for 0.11, 1.0, 1.1 and 2.0 and maintaining support for 2.1-2.8, 3.0-3.9 (17 releases total). Also, I believe the removal of the old exactly once support means that the minimum broker version for Kafka Streams with [exactly once becomes 2.5](https://issues.apache.org/jira/browse/KAFKA-16331?focusedCommentId=17876539&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17876539).

**Comment by ableegoldman** [comment]
@ijuma I suppose my concern is less about the number of versions dropped and more that, as far as I'm aware (and please correct me if this is wrong), users wouldn't have received any notice or even a warning log leading up to this?

I had this same concern about dropping eosv1 but ultimately felt that it was ok since it would still be possible to run some version of Streams (ie without eos) against older brokers.

I do want to clarify that I think this was/is the right decision, I'm just surprised it happened this quickly  -- it feels like a much bigger change than, for example, removing some niche rarely-used API -- but even for that we require both a deprecation notice and a full year before removal, so why not this?

**Comment by ableegoldman** [comment]
I was thinking we had an official compatibility guarantee for Kafka Streams but I took another look at the compatibility matrix page and it does only list compatibility for current/existing versions. So I can see the argument that it's not a formal contract and therefore doesn't require the usual deprecation/removal process.

Again, I'm not trying to block this, I just wished we did a better job of giving users a heads up. And by "we" I mean myself included, since I completely missed [KIP-896](https://cwiki.apache.org/confluence/display/KAFKA/KIP-896%3A+Remove+old+client+protocol+API+versions+in+Kafka+4.0) which is on me.

To be fair, we also haven't updated our compatibility matrix since 3.7, so it seems we(Streams) dropped the ball on updating docs all around 🙂

**Comment by ableegoldman** [comment]
FYI I opened a PR to fix/update the compatibility matrix and document the change in broker compatibility going forward. I had some questions for @ijuma  and @mjsax so please take a look when you have a minute. Thanks!

PR: [https://github.com/apache/kafka/pull/17906](https://github.com/apache/kafka/pull/17906)

**Comment by mjsax** [comment]
> I had this same concern about dropping eosv1 but ultimately felt that it was ok since it would still be possible to run some version of Streams (ie without eos) against older brokers.

We did officially deprecate EOSv1, and if people did still use it, we did log a warning about it getting removed. https://github.com/apache/kafka/blob/3.9/streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java#L1469-L1473 -- So there was more then enough time for people to adopt. Don't think this was rapid at all.

**Comment by ijuma** [comment]
@ableegoldman Note that we did announce this change publicly too:

https://kafka.apache.org/blog#apache_kafka_370_release_announcement

I agree that we could perhaps have done more on the client side to warn, but I focused on the broker side versus the client side because it seems highly unlikely that someone would use a Kafka 4.0 client against a Kafka 2.0 (or older) broker (Kafka 2.0 was released in 2018, no kafka vendor supports it at this point, there are a number of CVEs that apply to it, etc.).

The reverse where someone is using an ancient client against a newer broker is much more likely. That's also what we observed while talking to customers and users, they tend to be worried about the latter because clients are embedded in a large number of applications and are harder/slower to upgrade while I'm not aware of any of them raising concerns about the former.

**Comment by ableegoldman** [comment]
> We did officially deprecate EOSv1, and if people did still use it, we did log a warning about it getting removed. https://github.com/apache/kafka/blob/3.9/streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java#L1469-L1473 -- So there was more then enough time for people to adopt. Don't think this was rapid at all.

Right -- like I said, I  didn't have a problem with the eosv1 removal and didn't feel it was sudden/rapid at all. Partly also because of exactly this, ie that we took care to give users a clear warning ahead of time. I guess that's my main point in all of this: going forward we should think about how to properly warn users ahead of time

@ijuma I've definitely spoken with users who ran upgraded clients against older brokers, so they do exist, though statistically you may be right that the reverse is more common. And obviously this only applies to on-prem installations, which are certainly becoming less and less frequent but shouldn't be disregarded altogether.

(And yes, I've even. spoken with someone who complained about their Kafka cluster being stuck on 0.11 due to apprehension around upgrades -- but to be fair, this was like two years ago or more, so I really really hope they managed to upgrade since then)

**Comment by chia7712** [comment]
@mingdaoy Could you please remove the 0.10-related code from `kafka-run-class.sh`? We no longer run e2e under 0.10, so the check for 0.10 is unnecessary.

https://github.com/apache/kafka/blob/e1ba01d21440a1de011c116a0bb6933656200009/bin/kafka-run-class.sh#L119

File: bin/kafka-run-class.sh
Comment (mingdaoy) [inline bin/kafka-run-class.sh:126]: PTAL @chia7712

**Review by chia7712 — APPROVED** [review]

**Comment by chia7712** [comment]
I ran all Streams end-to-end tests with this PR, and `streams_eos_test.py` failed on both the trunk and this PR. @mjsax @ableegoldman any concerns about removing 0.10.x code? If not, I'd like to merge this PR today.

**Comment by mjsax** [comment]
I am personally ok to remove 0.10.0 and 0.10.1 -- both are _really_ old...

We just had a similar discussion about this on https://github.com/apache/kafka/pull/18055 -- not sure if @cadonna wants to chime in here?

**Comment by chia7712** [comment]
@cadonna Please feel free to raise any objections. This PR solely removes the 0.10 versions and can be easily reverted if necessary.

**Pull request merged**
